cpp/src/parquet/page_index_test.cc

```diff
@@ -0,0 +1,208 @@
+  PARQUET_THROW_NOT_OK(source->Close());
+
+  // Deserialize offset index.
+  auto properties = default_reader_properties();
+  std::unique_ptr<OffsetIndex> offset_index = OffsetIndex::Make(
+      buffer->data(), static_cast<uint32_t>(buffer->size()), properties);
+
+  // Verify only partial data as it contains 325 pages in total.
+  const size_t num_pages = 325;
+  const std::vector<size_t> page_indices = {0, 100, 200, 300};
+  const std::vector<PageLocation> page_locations = {
+      PageLocation{4, 109, 0}, PageLocation{11480, 133, 2244},
+      PageLocation{22980, 133, 4494}, PageLocation{34480, 133, 6744}};
+
+  ASSERT_EQ(num_pages, offset_index->page_locations().size());
+  for (size_t i = 0; i < page_indices.size(); ++i) {
+    size_t page_id = page_indices.at(i);
+    const auto& read_page_location = offset_index->page_locations().at(page_id);
+    const auto& expected_page_location = page_locations.at(i);
+    ASSERT_EQ(expected_page_location.offset, read_page_location.offset);
+    ASSERT_EQ(expected_page_location.compressed_page_size,
+              read_page_location.compressed_page_size);
+    ASSERT_EQ(expected_page_location.first_row_index, read_page_location.first_row_index);
+  }
+}
+
+template <typename DType, typename T = typename DType::c_type>
+void TestReadTypedColumnIndex(
+    int column_id, size_t num_pages, BoundaryOrder::type boundary_order,
+    const std::vector<size_t>& page_indices, const std::vector<bool>& null_pages,
+    const std::vector<T>& min_values, const std::vector<T>& max_values,
+    bool has_null_counts = false, const std::vector<int64_t>& null_counts = {}) {
+  std::string dir_string(parquet::test::get_data_dir());
+  std::string path = dir_string + "/alltypes_tiny_pages.parquet";
+  auto reader = ParquetFileReader::OpenFile(path, false);
+  auto file_metadata = reader->metadata();
+
+  // Get column index location to a specific column chunk.
+  const int row_group_id = 0;
+  ASSERT_LT(row_group_id, file_metadata->num_row_groups());
+  ASSERT_LT(column_id, file_metadata->num_columns());
+  auto index_location = file_metadata->RowGroup(row_group_id)
+                            ->ColumnChunk(column_id)
+                            ->GetColumnIndexLocation();
+  ASSERT_TRUE(index_location.has_value());
+
+  // Read serialized column index from the file.
+  std::shared_ptr<::arrow::io::RandomAccessFile> source;
+  PARQUET_ASSIGN_OR_THROW(source, ::arrow::io::ReadableFile::Open(path));
+  PARQUET_ASSIGN_OR_THROW(auto buffer,
+                          source->ReadAt(index_location->offset, index_location->length));
+  PARQUET_THROW_NOT_OK(source->Close());
+
+  // Deserialize column index.
+  auto properties = default_reader_properties();
+  auto descr = file_metadata->schema()->Column(column_id);
+  std::unique_ptr<ColumnIndex> column_index = ColumnIndex::Make(
+      *descr, buffer->data(), static_cast<uint32_t>(buffer->size()), properties);
```

Comment:
Can we perhaps add an API to `ColumnReader` or `RowGroupReader` that would fetch the `ColumnIndex` of a column?

(if so, then perhaps do the same for `OffsetIndex`)

Reply:
That is exactly in the plan of next patch. :)